onnxruntime/core/providers/cpu/ml/tree_ensemble_aggregator.h

```diff
@@ -64,6 +64,15 @@ enum MissingTrack : uint8_t {
   kFalse = 0
 };
 
+template <typename T>
+struct TreeNodeElement;
+
+template <typename T>
+union PtrOrWeightPos {
+  TreeNodeElement<T>* ptr;
```

Comment:
The current code was made to reduce the memory footprint. The size of `TreeNodeElement<T>*` is 8 bytes so it will add 8 = 4*2 bytes to every node.

Reply:
I've addressed this in the latest commit. `TreeNodeElement` has the exact same memory footprint. By laying out the nodes in preordered fashion we can guarantee the false branch is the next element in nodes_, meaning we don't need to store it and can reorganise things.